### PR TITLE
Add unstable Iterator::copied()

### PR DESCRIPTION
**Pull request by sugar700**
Initially suggested at https://github.com/bluss/rust-itertools/pull/289, however the maintainers of itertools suggested this may be better of in a standard library.

The intent of `copied` is to avoid accidentally cloning iterator elements after doing a code refactoring which causes a structure to be no longer `Copy`. This is a relatively common pattern, as it can be seen by calling `rg --pcre2 '[.]map[(][|](?:(\w+)[|] [*]\1|&(\w+)[|] \2)[)]'` on Rust main repository. Additionally, many uses of `cloned` actually want to simply `Copy`, and changing something to be no longer copyable may introduce unnoticeable performance penalty.

Also, this makes sense because the standard library includes `[T].copy_from_slice` to pair with `[T].clone_from_slice`.

This also adds `Option::copied`, because it makes sense to pair it with `Iterator::copied`. I don't think this feature is particularly important, but it makes sense to update `Option` along with `Iterator` for consistency.

### CODE REVIEW

File: src/libcore/iter/mod.rs [hunk 1]
```diff
@@ -501,6 +501,104 @@ impl<I> FusedIterator for Rev<I>
...
+        self.it.fold(init, move |acc, &elt| f(acc, elt))
+    }
+}
+
+#[unstable(feature = "iter_copied", issue = "0")]
+impl<'a, I, T: 'a> DoubleEndedIterator for Copied<I>
+    where I: DoubleEndedIterator<Item=&'a T>, T: Copy
+{
+    fn next_back(&mut self) -> Option<T> {
+        self.it.next_back().cloned()
```
Comment (bluss) [inline src/libcore/iter/mod.rs]: should be copied, same in methods below

File: src/libcore/iter/mod.rs [hunk 2]
```diff
@@ -501,6 +501,104 @@ impl<I> FusedIterator for Rev<I>
...
+#[doc(hidden)]
+unsafe impl<'a, I, T: 'a> TrustedRandomAccess for Copied<I>
+    where I: TrustedRandomAccess<Item=&'a T>, T: Copy
+{
+    unsafe fn get_unchecked(&mut self, i: usize) -> Self::Item {
+        *self.it.get_unchecked(i)
+    }
+
+    #[inline]
+    fn may_have_side_effect() -> bool { false }
```
Comment (bluss) [inline src/libcore/iter/mod.rs]: This should be `I::may_have_side_effect()`, we don't add side effects but propagate the ones from the adapted iterator

File: src/libcore/iter/mod.rs [hunk 3]
```diff
@@ -501,6 +501,104 @@ impl<I> FusedIterator for Rev<I>
...
+    }
+
+    fn try_fold<B, F, R>(&mut self, init: B, mut f: F) -> R where
+        Self: Sized, F: FnMut(B, Self::Item) -> R, R: Try<Ok=B>
+    {
+        self.it.try_fold(init, move |acc, &elt| f(acc, elt))
+    }
+
+    fn fold<Acc, F>(self, init: Acc, mut f: F) -> Acc
+        where F: FnMut(Acc, Self::Item) -> Acc,
```
[the remark this answers is not in the document]
Comment (xfix) [reply src/libcore/iter/mod.rs]: It's consistently so in this file, with different indentation for `try_fold` and `fold`. It probably should be fixed, but seems out of scope for this change.

[the remark this answers is not in the document]
Comment (bluss) [reply src/libcore/iter/mod.rs]: uh that makes sense I guess :slightly_smiling_face:

File: src/libcore/iter/mod.rs [hunk 4]
```diff
@@ -501,6 +501,106 @@ impl<I> FusedIterator for Rev<I>
...
+    where I: DoubleEndedIterator<Item=&'a T>, T: Copy
+{
+    fn next_back(&mut self) -> Option<T> {
+        self.it.next_back().copied()
+    }
+
+    fn try_rfold<B, F, R>(&mut self, init: B, mut f: F) -> R where
+        Self: Sized, F: FnMut(B, Self::Item) -> R, R: Try<Ok=B>
+    {
+        self.it.try_rfold(init, move |acc, elt| f(acc, elt.clone()))
```
Comment (bluss) [inline src/libcore/iter/mod.rs]: This one should copy like `try_fold`

File: src/libcore/iter/mod.rs [hunk 5]
```diff
@@ -501,6 +501,106 @@ impl<I> FusedIterator for Rev<I>
...
+    }
+
+    fn rfold<Acc, F>(self, init: Acc, mut f: F) -> Acc
+        where F: FnMut(Acc, Self::Item) -> Acc,
+    {
+        self.it.rfold(init, move |acc, elt| f(acc, elt.clone()))
```
Comment (bluss) [inline src/libcore/iter/mod.rs]: This one should copy like `fold`.

File: src/libcore/iter/mod.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (Lucretiel) [reply src/libcore/iter/mod.rs]: I was thinking that it might be good to (at some point) run rustfmt on the whole standard library

**Comment by Lucretiel** [comment]
Quick note: there's currently a clippy lint that suggests replacing `iter.map(|x| *x)` with `iter.cloned()`. Would probably be worth updating that lint once this API stabilizes.

File: src/libcore/iter/mod.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (bluss) [reply src/libcore/iter/mod.rs]: That's coming

**Comment by SimonSapin** [comment]
For what it’s worth `copy_from_slice` is not "just" `clone_from_slice` with `*x` instead of `x.clone()`, it is implemented as a single call to `ptr::copy_nonoverlapping`.

**Comment by xfix** [comment]
While true, a similar argument applies to `Cloned`, as it actually has a specialization for `T: Copy`.

**Comment by shepmaster** [comment]
The feature makes sense to me. @bluss has done the existing superb job of review, so I think it's better if they get control of the big button ;-)

r? @bluss

**Comment by bluss** [comment]
Technically the PR is fine. Thoughts on including this in std?

**Comment by Centril** [comment]
(My personal opinion is that this would be useful as a means of ensuring "cheapness")

Randomly re-assigning to someone on T-libs for final triage.

r? @SimonSapin

**Comment by SimonSapin** [comment]
Diff looks good. r=me with a tracking issue filed and referenced in `#[unstable]` attributes.

I believe we tend to default to accepting new unstable APIs, and revisit them when stabilization is proposed.

**Comment by xfix** [comment]
@SimonSapin created tracking issues.

**Pull request merged**